### PR TITLE
[2/3] Remove data stuff and use Ray Data.

### PR DESCRIPTION
**Pull request by coledie**
- [x] Forked repo is synced with upstream -> github shows no code delta outside of the desired.
    https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork
- [x] Tests are passing? https://github.com/microsoft/DeepGNN/blob/main/CONTRIBUTING.md#run-tests
- [x] Changelog and documentation updated.
- [x] PR is labeled using the label menu on the right side.

New Behavior
----------------
* Remove backends, TorchDeepGNNDataset and samplers.
* Examples now use Ray Data.

### CODE REVIEW

File: examples/pytorch/gat/ray_util.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
+"""Baseline Ray Trainer."""
```
Comment (coledie) [inline examples/pytorch/gat/ray_util.py:1]: TODO remove, revert to migrate torch version.

File: examples/pytorch/graphsage/main.py [hunk 2]
```diff
@@ -170,7 +141,7 @@ def train_func(config: Dict):
 
 def run_ray(init_dataset_fn, **kwargs):
     """Run ray trainer."""
-    ray.init(num_cpus=3)
+    ray.init(num_cpus=4)
```
Comment (mortid0) [inline examples/pytorch/graphsage/main.py:144]: We probably should autodetect cpus..

File: examples/pytorch/graphsage/model.py [hunk 3]
```diff
@@ -93,9 +93,9 @@ def query(self, graph: Graph, inputs: np.ndarray) -> dict:
             n0_out, np.array([[self.feature_idx, self.feature_dim]]), self.feature_type
         )
 
-        context["x0"] = x0
-        context["out_1"] = n1_out.shape[0]  # Number of output nodes of layer 1
-        context["out_2"] = n2_out.shape[0]  # Number of output nodes of layer 2
+        context["x0"] = x0.reshape((context["inputs"].shape[0], -1, self.feature_dim))
+        context["out_1"] =  np.array([n1_out.shape[0]] * context["inputs"].shape[0])  # Number of output nodes of layer 1
```
Comment (mortid0) [inline examples/pytorch/graphsage/model.py:97]: why do you need to create arrays instead of passing shapes?

File: examples/pytorch/hetgnn/main.py [hunk 4]
```diff
@@ -23,11 +21,12 @@
 from ray.train.torch import TorchTrainer
 from ray.air import session
 from ray.air.config import ScalingConfig
+from ray.data import DatasetPipeline
+import pyarrow as pa
```
Comment (mortid0) [inline examples/pytorch/hetgnn/main.py:25]: where do you use pyarrow?

File: examples/pytorch/hetgnn/sampler.py [hunk 5]
```diff
@@ -23,7 +23,7 @@ def __init__(
         sample_files: str = "",
     ):
         """Initialize sampler."""
-        super().__init__(batch_size, epochs=1, shuffle=False)
+        #super().__init__(batch_size, epochs=1, shuffle=False)
```
Comment (mortid0) [inline examples/pytorch/hetgnn/sampler.py:26]: remove commented out code?

File: examples/pytorch/knowledgegraph/main.py [hunk 6]
```diff
@@ -62,31 +58,27 @@ def train_func(config: Dict):
...
+    address = "localhost:9999"
+    s = Server(address, args.data_dir, 0, len(args.partitions))
+    g = DistributedClient([address])
+    # NOTE: See https://deepgnn.readthedocs.io/en/latest/graph_engine/dataset.html
+    #       for how to use a different sampler
+    max_id = g.node_count(args.node_type) if args.max_id in [-1, None] else args.max_id
+    dataset = ray.data.range(max_id).repartition(max_id // args.batch_size)
+    pipe = dataset.window(blocks_per_window=4).repeat(args.num_epochs)
+    def transform_batch(idx: list) -> dict:
+        # If get Ray error with return shape, use deepgnn.graph_engine.util.serialize/deserialize
```
Comment (mortid0) [inline examples/pytorch/knowledgegraph/main.py:70]: when does this error happen and can we avoid it?

File: examples/pytorch/knowledgegraph/main.py [hunk 7]
```diff
@@ -62,31 +58,27 @@ def train_func(config: Dict):
...
+        # in your query and forward function
+        return model.query(g, np.array(idx))  # TODO Update to your query function
```
Comment (mortid0) [inline examples/pytorch/knowledgegraph/main.py:72]: re: TODO in this PR?

File: src/python/deepgnn/graph_engine/utils.py [hunk 8]
```diff
@@ -97,3 +81,37 @@ def define_param_graph_engine(parser: argparse.ArgumentParser):
         default="",
         help="Directory to synchronize start of graph engine and workers.",
     )
+
+
+def serialize(inputs, batch_size):
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/utils.py:86]: where is documentation and type hints?

File: examples/pytorch/graphsage/main.py (hunk 2 above)
Comment (coledie) [reply examples/pytorch/graphsage/main.py:156]: This is only necessary because CI pipeline does not have enough cpus.

File: examples/pytorch/graphsage/model.py (hunk 3 above)
Comment (coledie) [reply examples/pytorch/graphsage/model.py]: I need to convert from an int to an array with size =batch_size in order to serialze with ray

File: examples/pytorch/knowledgegraph/main.py (hunk 7 above)
Comment (coledie) [reply examples/pytorch/knowledgegraph/main.py]: This was supposed to be docs for the user who is migrating.
